### DISCUSSION TITLE
How to correctly compute the area of projection on a plane?

### DISCUSSION THREAD

**Discussion by subodh-malgonde** [pyvista/pyvista#5211] [Q&A]
I am projecting a `PolyData` object onto a plane. Then I want to compute the surface area of the projection.

Lets take a simple cube as an example. Lets project it onto a plane which is parallel to the XY plane.

```
import pyvista as pv

cube = pv.Cube()

origin = cube.center
origin[-1] -= 2.0 # Plane is 2 meters below the origin

projected = cube.project_points_to_plane(origin=origin, normal=[0,0,1], inplace=False)

plotter = pv.Plotter()
plotter.add_mesh(cube)
plotter.add_mesh(projected, color='gray')
plotter.show_axes()
plotter.show()
```

This is how it looks like:

![cube_projection](https://user-images.githubusercontent.com/5463437/114843419-6dd39d80-9df7-11eb-9ad6-5043d0ed56fb.png)

To compute the area of the projection I did this:

```
projected.compute_cell_sizes()['Area'].sum()
```

The output is `2.0`. I had expected the output to be `1.0`.

On inspecting the the cell areas array, it looks like all the 6 faces of the cube are preserved. With zero area for the 4 faces perpendicular to the plane.

```
array([0., 0., 0., 0., 1., 1.])
```

## My question
Now in this case I can simple divide the result by `2` to get the actual surface area. But what if it is a more complicated geometry, how do I accurately compute the surface area of the projection?

**Comment by tkoyama010** [comment]
@subodh-malgonde Thank you for your question. As you check it is because this projected geometory also have top and bottom surface. If you want to to calculate only 1 side. Plese remove other side using [clip](https://docs.pyvista.org/core/filters.html#pyvista.DataSetFilters.clip) or [clip_box](https://docs.pyvista.org/core/filters.html#pyvista.DataSetFilters.clip_box) method.

**Comment by MattTheCuber** [comment]
> @subodh-malgonde Thank you for your question. As you check it is because this projected geometory also have top and bottom surface. If you want to to calculate only 1 side. Plese remove other side using [clip](https://docs.pyvista.org/core/filters.html#pyvista.DataSetFilters.clip) or [clip_box](https://docs.pyvista.org/core/filters.html#pyvista.DataSetFilters.clip_box) method.

Sorry this is old, but what about for much more complex shapes such as meshes with many faces and concave details.

Edit: Found a [similar issue](https://github.com/pyvista/pyvista/issues/2842)

**Reply by darikg** [reply]
@MattTheCuber use a 2d geometry library like shapely to find the union of all the projected triangles and take the area of that. (Don't even need to worry about the normal of the projected faces.)

```python
import pyvista
import shapely

mesh = pyvista.Superquadric(
    phi_roundness=5,
    theta_roundness=3,
).project_points_to_plane(origin=(0, 0, 0), normal=(0, 0, 1))

merged = shapely.union_all([
    shapely.Polygon(mesh.points[tri, :2])
    for tri in mesh.triangulate().regular_faces
])

merged.area
```

if you're not projecting to the xy-plane, will need an extra transform step to rotate the projected points into that plane so you can pass the xy-coords to shapely.

**Reply by MattTheCuber** [reply]
@darikg 
Thank you very much for this information. We have been searching for a good way to compute this information and the code you provided seems to work great! I was digging into the shapely code to try to find what they are doing in the `union_all` function and it appears that it just wrapping a function from [GEOS](https://github.com/libgeos/geos) library. I know its a long shot, but do you know what method is being done to compute the union? I saw an options for the function called `grid_size` which makes me think its using rasterization but I am not sure.

**Reply by darikg** [reply]
It definitely doesn't do rasterization but beyond that I have no idea! My understanding is that `unary_union` is more efficient than calling `union` on all of the triangles one at a time. Maybe it constructs an R-tree or other fast spatial indexing structure first to do that?

**Reply by MattTheCuber** [reply]
> It definitely doesn't do rasterization but beyond that I have no idea! My understanding is that `unary_union` is more efficient than calling `union` on all of the triangles one at a time. Maybe it constructs an R-tree or other fast spatial indexing structure first to do that?

Great, I saw an issue while researching saying that the library has been in use for over a decade now which proves it have stood the test of time.

Do you think it is safe to assume that the function is computing the actual areas and the accuracy is 100% as compared to rasterization which is never truly 100%?
